proxy.go

```diff
@@ -0,0 +1,93 @@
+//
+// Copyright (c) 2016 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package virtcontainers
+
+import (
+	"fmt"
+	"os"
+)
+
+// ProxyType describes a proxy type.
+type ProxyType string
+
+const (
+	// RawHyperstartProxy is the raw hyperstart proxy.
+	RawHyperstartProxy ProxyType = "rawHyperstartProxy"
+
+	// HyperstartProxy is the hyperstart proxy.
+	HyperstartProxy ProxyType = "hyperstartProxy"
+)
+
+// Set sets a proxy type based on the input string.
+func (pType *ProxyType) Set(value string) error {
+	switch value {
+	case "rawHyperstartProxy":
+		*pType = RawHyperstartProxy
+		return nil
+	case "hyperstartProxy":
+		*pType = HyperstartProxy
+		return nil
+	default:
+		return fmt.Errorf("Unknown proxy type %s", value)
+	}
+}
+
+// String converts a proxy type to a string.
+func (pType *ProxyType) String() string {
+	switch *pType {
+	case RawHyperstartProxy:
+		return string(RawHyperstartProxy)
+	case HyperstartProxy:
+		return string(HyperstartProxy)
+	default:
+		return ""
+	}
+}
+
+// newProxy returns a proxy from a proxy type.
+func newProxy(pType ProxyType) (proxy, error) {
+	switch pType {
+	case RawHyperstartProxy:
+		return &rawHyperstartProxy{}, nil
+	case HyperstartProxy:
+		return &hyperstartProxy{}, nil
+	default:
+		return &rawHyperstartProxy{}, nil
+	}
+}
+
+// proxy is the virtcontainers proxy interface.
+type proxy interface {
+	// connect connects the proxy to the agent inside the VM.
+	connect(pod Pod) error
+
+	// disconnect disconnects the proxy from the agent inside the VM.
+	disconnect(podID string) error
+
+	// attach attaches current proxy connection to a previously registered VM.
+	attach(pod Pod) error
+
+	// detach detaches current proxy connection from a previously registered VM.
+	detach() error
+
+	// allocateIO asks the proxy to allocate IO stream sequence numbers and to
+	// route the IO from hyperstart to the right socket.
+	allocateIO(nStreams int) (uint64, *os.File, error)
+
+	// sendCmd sends a command to the agent inside the VM through the proxy.
+	sendCmd(cmd string, message interface{}) error
```

Comment:
Here we're assuming that none of the commands will ever send us an answer other than an error. Although this is true for `cc-proxy`, I'd like this one to be a little more generic and allow for returning an answer.